File: leetcode/problems/783.minimum-distance-between-bst-nodes.py
#
# @lc app=leetcode id=783 lang=python3
#
# [783] Minimum Distance Between BST Nodes
#


# @lc code=start
# Definition for a binary tree node.
import math


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right


class Solution:
    prev = -math.inf
    min_diff = math.inf

    def minDiffInBST(self, root: Optional[TreeNode]) -> int:
        if root.left:
            self.minDiffInBST(root.left)

        self.min_diff = min(self.min_diff, root.val - self.prev)
        self.prev = root.val

        if root.right:
            self.minDiffInBST(root.right)

        return self.min_diff


# @lc code=end

File: leetcode/problems/783.minimum-distance-between-bst-nodes.java
/*
 * @lc app=leetcode id=783 lang=java
 *
 * [783] Minimum Distance Between BST Nodes
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode() {
    }

    TreeNode(int val) {
        this.val = val;
    }

    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    private int prev = Integer.MIN_VALUE;
    private int min_diff = Integer.MAX_VALUE;

    public int minDiffInBST(TreeNode root) {
        if (root.left != null) {
            this.minDiffInBST(root.left);
        }
        if (prev != Integer.MIN_VALUE) {
            min_diff = Math.min(min_diff, root.val - prev);
        }
        prev = root.val;

        if (root.right != null) {
            this.minDiffInBST(root.right);
        }

        return min_diff;
    }
}
// @lc code=end
